File: Angular-With-Boot/Running Boot/src/main/java/com/mrahmed/HRandPayrollManagementSystem/service/AdvanceSalaryService.java
package com.mrahmed.HRandPayrollManagementSystem.service;

import com.mrahmed.HRandPayrollManagementSystem.entity.AdvanceSalary;
import com.mrahmed.HRandPayrollManagementSystem.entity.Month;
import com.mrahmed.HRandPayrollManagementSystem.repository.AdvanceSalaryRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

@Service
public class AdvanceSalaryService {


    @Autowired
    private AdvanceSalaryRepository advanceSalaryRepository;

    // Save new advance salary record
    public AdvanceSalary saveAdvanceSalary(AdvanceSalary advanceSalary) {
        return advanceSalaryRepository.save(advanceSalary);
    }

    // Update existing advance salary record
    public AdvanceSalary updateAdvanceSalary(AdvanceSalary advanceSalary) {
        if (advanceSalary.getId() != 0 && advanceSalaryRepository.existsById(advanceSalary.getId())) {
            return advanceSalaryRepository.save(advanceSalary);
        }
        throw new IllegalArgumentException("AdvanceSalary with ID " + advanceSalary.getId() + " does not exist.");
    }

    // Delete advance salary record by ID
    public void deleteAdvanceSalary(Long id) {
        if (advanceSalaryRepository.existsById(id)) {
            advanceSalaryRepository.deleteById(id);
        } else {
            throw new IllegalArgumentException("AdvanceSalary with ID " + id + " does not exist.");
        }
    }

    // Find advance salary record by ID
    public Optional<AdvanceSalary> getAdvanceSalaryById(Long id) {
        return advanceSalaryRepository.findById(id);
    }

    // Find advance salaries by user and year
    public List<AdvanceSalary> getAdvanceSalariesByUserAndYear(Long userId, int year) {
        return advanceSalaryRepository.findByUserIdAndYear(userId, year);
    }

    // Find advance salaries by user, year, and month
    public List<AdvanceSalary> getAdvanceSalariesByUserYearAndMonth(Long userId, int year, Month month) {
        return advanceSalaryRepository.findByUserIdYearAndMonth(userId, year, month);
    }

    // Calculate total advance salary for a user in a specific year
    public BigDecimal getTotalAdvanceSalaryByUserAndYear(Long userId, int year) {
        return advanceSalaryRepository.getTotalAdvanceSalaryByUserAndYear(userId, year);
    }

    // Find advance salaries within a specific date range
    public List<AdvanceSalary> getAdvanceSalariesByDateRange(LocalDateTime startDate, LocalDateTime endDate) {
        return advanceSalaryRepository.findByDateRange(startDate, endDate);
    }

    // Find advance salaries for a specific month and year
    public List<AdvanceSalary> getAdvanceSalariesByMonthAndYear(Month month, int year) {
        return advanceSalaryRepository.findByMonthAndYear(month, year);
    }

    // Find latest advance salary record for a user
    public List<AdvanceSalary> getLatestAdvanceSalaryByUser(Long userId) {
        return advanceSalaryRepository.findLatestAdvanceSalaryByUser(userId);
    }

}

File: Angular-With-Boot/Running Angular v7/src/app/services/advancesalary.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { AdvanceSalaryModel } from '../models/advance-salary.model';
import { environment } from '../../environments/environment';

@Injectable({
  providedIn: 'root',
})
export class AdvanceSalaryService {
  private baseUrl = `${environment.apiUrl}/advanceSalaries`;

  constructor(private http: HttpClient) {}

  // Create a new advance salary
  createAdvanceSalary(
    advanceSalary: AdvanceSalaryModel
  ): Observable<AdvanceSalaryModel> {
    return this.http.post<AdvanceSalaryModel>(
      `${this.baseUrl}/create`,
      advanceSalary
    );
  }

  // Update an existing advance salary
  updateAdvanceSalary(
    id: number,
    advanceSalary: AdvanceSalaryModel
  ): Observable<AdvanceSalaryModel> {
    return this.http.put<AdvanceSalaryModel>(
      `${this.baseUrl}/update/${id}`,
      advanceSalary
    );
  }

  // Delete an advance salary by ID
  deleteAdvanceSalary(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/delete/${id}`);
  }

  // Get an advance salary by ID
  getAdvanceSalaryById(id: number): Observable<AdvanceSalaryModel> {
    return this.http.get<AdvanceSalaryModel>(`${this.baseUrl}/find/${id}`);
  }

  // Get advance salaries by user and year
  getAdvanceSalariesByUserAndYear(
    userId: number,
    year: number
  ): Observable<AdvanceSalaryModel[]> {
    return this.http.get<AdvanceSalaryModel[]>(
      `${this.baseUrl}/user/${userId}/year/${year}`
    );
  }

  // Get advance salaries by user, year, and month
  getAdvanceSalariesByUserYearAndMonth(
    userId: number,
    year: number,
    month: string
  ): Observable<AdvanceSalaryModel[]> {
    return this.http.get<AdvanceSalaryModel[]>(
      `${this.baseUrl}/user/${userId}/year/${year}/month/${month}`
    );
  }

  // Get total advance salary for a user in a specific year
  getTotalAdvanceSalaryByUserAndYear(
    userId: number,
    year: number
  ): Observable<number> {
    return this.http.get<number>(
      `${this.baseUrl}/user/${userId}/total-year/${year}`
    );
  }

  // Get advance salaries within a specific date range
  getAdvanceSalariesByDateRange(
    startDate: Date,
    endDate: Date
  ): Observable<AdvanceSalaryModel[]> {
    const params = new HttpParams()
      .set('startDate', startDate.toISOString())
      .set('endDate', endDate.toISOString());
    return this.http.get<AdvanceSalaryModel[]>(`${this.baseUrl}/date-range`, {
      params,
    });
  }

  // Get advance salaries for a specific month and year
  getAdvanceSalariesByMonthAndYear(
    month: string,
    year: number
  ): Observable<AdvanceSalaryModel[]> {
    return this.http.get<AdvanceSalaryModel[]>(
      `${this.baseUrl}/month/${month}/year/${year}`
    );
  }

  // Get the latest advance salary record for a user
  getLatestAdvanceSalaryByUser(
    userId: number
  ): Observable<AdvanceSalaryModel[]> {
    return this.http.get<AdvanceSalaryModel[]>(
      `${this.baseUrl}/latest/user/${userId}`
    );
  }
}
